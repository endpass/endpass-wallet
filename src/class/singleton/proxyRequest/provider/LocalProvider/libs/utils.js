import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
