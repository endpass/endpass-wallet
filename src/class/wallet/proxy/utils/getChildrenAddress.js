export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
