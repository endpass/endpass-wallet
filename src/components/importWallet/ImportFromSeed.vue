<template>
  <v-form @submit="submitAddWallet">
    <v-input
      id="hdkeySeed"
      key="hdkeyPhraseUnique"
      v-model="key"
      label="Seed phrase"
      name="hdkeyPhrase"
      validator="required|seed_phrase"
      data-vv-as="seed phrase"
      aria-describedby="hdkeyPhrase"
      placeholder="Seed phrase"
      required
      @input="handleInput"
    />
    <v-password
      v-model="password"
      label="Wallet password"
      name="password"
      validator="required|min:8"
      data-vv-as="password"
      aria-describedby="jsonKeystorePassword"
      placeholder="wallet password"
      required
    />
    <v-button
      :loading="isCreating"
      class-name="is-primary is-cta"
    >
      Import
    </v-button>
  </v-form>
</template>

<script>
import { mapActions } from 'vuex';
import VForm from '@/components/ui/form/VForm.vue';
import VInput from '@/components/ui/form/VInput.vue';
import VPassword from '@/components/ui/form/VPassword.vue';
import VButton from '@/components/ui/form/VButton.vue';

export default {
  name: 'ImportFromSeed',
  data: () => ({
    isCreating: false,
    key: '',
    password: '',
  }),
  methods: {
    ...mapActions('accounts', ['addMultiHdWallet']),
    async submitAddWallet() {
      this.isCreating = true;

      await new Promise(res => setTimeout(res, 20));

      try {
        this.addMultiHdWallet({
          key: this.key,
          password: this.password,
        });
        this.$router.push('/');
      } catch (e) {
        this.errors.add({
          field: 'hdkeyPhrase',
          msg: 'Seed phrase is invalid',
          id: 'wrongPhrase',
        });
        console.error(e);
      }

      this.isCreating = false;
    },
    handleInput() {
      this.errors.removeById('wrongPhrase');
    },
  },
  components: {
    VForm,
    VInput,
    VPassword,
    VButton,
  },
};
</script>

<style lang="scss">
</style>
