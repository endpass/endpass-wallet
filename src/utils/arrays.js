export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
