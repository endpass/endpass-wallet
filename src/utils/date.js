import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

/**
 * Setting up dayjs globally
 */
dayjs.extend(relativeTime);

export const formateDate = date => dayjs(date).format('YYYY-MM-DD H:mm');

export const fromNow = date => dayjs(date).fromNow();

export default {
  formateDate,
  fromNow,
};
