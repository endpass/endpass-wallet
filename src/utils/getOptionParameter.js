export default function getOptionParameter(item, value) {
  return item instanceof Object ? item[value] : item;
}
