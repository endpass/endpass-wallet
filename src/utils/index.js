import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
