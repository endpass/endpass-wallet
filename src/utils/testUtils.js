import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
