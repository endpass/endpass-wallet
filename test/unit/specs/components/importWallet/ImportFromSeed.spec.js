import { shallow, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import VueRouter from 'vue-router';

import { generateStubs } from '@/utils/testUtils';

import ImportFromSeed from '@/components/importWallet/ImportFromSeed';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);
localVue.use(VeeValidate);

jest.useFakeTimers();
describe('ImportFromSeed', () => {
  let wrapper, actions, router;
  beforeEach(() => {
    actions = {
      addMultiHdWallet: jest.fn(),
    };
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions,
        },
      },
    };
    router = new VueRouter();
    const store = new Vuex.Store(storeOptions);
    wrapper = shallow(ImportFromSeed, {
      localVue,
      store,
      router,
      stubs: generateStubs(ImportFromSeed),
    });
  });
  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('ImportFromSeed');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('addWallet', () => {
      it('should call vuex addMultiHdWallet with correct arguments', done => {
        const data = {
          key: 'kek',
          password: 'kek',
        };

        expect.assertions(1);

        wrapper.setData(data);
        wrapper.vm.submitAddWallet().then(() => {
          expect(actions.addMultiHdWallet).toBeCalledWith(
            expect.any(Object),
            data,
            undefined,
          );
          done();
        });
        jest.runAllTimers();
      });

      it('should redirect to root after successful wallet creation', done => {
        expect.assertions(2);
        router.push('/kek');
        expect(router.currentRoute.fullPath).toBe('/kek');
        wrapper.vm.submitAddWallet().then(() => {
          expect(router.currentRoute.fullPath).toBe('/');
          done();
        });
        jest.runAllTimers();
      });

      it('should toggle isCreating before and after wallet creation', done => {
        expect.assertions(4);

        wrapper.vm.submitAddWallet().then(() => {
          expect(actions.addMultiHdWallet).toBeCalled();
          expect(wrapper.vm.isCreating).toBe(false);
          done();
        });

        expect(actions.addMultiHdWallet).not.toBeCalled();
        expect(wrapper.vm.isCreating).toBe(true);
        jest.runAllTimers();
      });

      it('should add error to field if failed to create wallet', done => {
        actions.addMultiHdWallet.mockImplementationOnce(() => {
          throw new Error();
        });
        wrapper.vm.submitAddWallet().catch(() => {
          expect(wrapper.vm.errors.items[0]).toEqual({
            field: 'hdkeyPhrase',
            msg: 'Seed phrase is invalid',
            id: 'wrongPhrase',
            // vee validate added field
            scope: null,
          });
          done();
        });
        jest.runAllTimers();

        expect.assertions(1);
      });
    });

    describe('handleInput', () => {
      it('should clear error with wrongAddress id', () => {
        wrapper.vm.errors.add({
          field: 'hdkeyPhrase',
          msg: 'Seed phrase is invalid',
          id: 'wrongPhrase',
        });
        expect(wrapper.vm.errors.has('hdkeyPhrase')).toBe(true);
        wrapper.vm.handleInput();
        expect(wrapper.vm.errors.has('hdkeyPhrase')).toBe(false);
      });
    });
  });
});
