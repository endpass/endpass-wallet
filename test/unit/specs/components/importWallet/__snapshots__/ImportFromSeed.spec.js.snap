// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportFromSeed render should render initial state of the component 1`] = `
<div>
  <v-form
    v-on:submit="bound togglePasswordModal"
  >
    <v-input
      aria-describedby="hdkeyPhrase"
      data-vv-as="seed phrase"
      id="hdkeySeed"
      label="Seed phrase"
      name="hdkeyPhrase"
      placeholder="Seed phrase"
      required=""
      v-model="key"
      v-on:input="bound handleInput"
      validator="required|seed_phrase"
    />
     
    <v-button
      class-name="is-primary is-cta"
    >
      
      Import
    
    </v-button>
  </v-form>
   
  <!---->
</div>
`;
